experiments/benchmarks/utils/integrands/madgraph.py

```diff
@@ -0,0 +1,223 @@
+        pdf: boolean
+            If True, the beams are assuemed to be proton beams and the PDFs are included in the calculation.
+        delr_cut: float or torch.Tensor
+            The minimum angular distance between each pair of the outgoing particles. 
+        pt_cut: float or torch.Tensor
+            The minimal transversal momentum of each outgoing particle.
+        rap_maxcut: float or torch.Tensor
+            The maximal rapidity of each outgoing particle. -1 means no cut.
+        pdf_type: String
+            The name of the PDF set to be used by LHAPDF.
+        pdf_dir: String
+            The directory whre the PDF data sets are stored.
+        lhapdf_dir: String
+            The path to the LHAPDF shared-object file.
+        process: String
+            The name given by MadGraph5_aMC@NLO to the subprocess.
+        device: torch.device
+            Default device where the parameters are stored
+        """
+        
+        logger.debug(process)
+        self.pdf=pdf
+        self.pdf_obj=None
+        
+        #Try to import LHAPDF. If this is not possible, PDFs are not included
+        if self.pdf and lhapdf_dir==None:
+            logger.debug("The directory of the LHAPDF module was not given. PDFs are not included")
+            self.pdf=False
+        if self.pdf:
+            print(lhapdf_dir)
+            if lhapdf_dir not in sys.path:
+                sys.path.append(lhapdf_dir)
+            try:
+                import lhapdf
+                self.pdf_obj = lhapdf.mkPDF(pdf_type, 0)
+            except Exception as e:
+                logger.error("LHAPDF could not be imported")
+                logger.debug(e)
+                self.pdf=False
+                
+        BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+        
+        
+        #Try to import the matrix element shared-object file
+        if BASE_DIR+"/integrands/mg/"+process not in sys.path:
+            sys.path.append(BASE_DIR+"/integrands/mg/"+process)
+        try:
+            import matrix2py
+        except:
+            logger.debug("The matrix elements could not be imported")
+            
+        self.process=process
+        self.E_cm=sanitize_variable(e_cm,device)
+        self.delR_cut=sanitize_variable(delr_cut,device)
+        self.pT_cut=sanitize_variable(pt_cut,device)
+        self.rap_maxcut=sanitize_variable(rap_maxcut,device)
+        self.default_device=device
+
+        #the mapping between particles and pdg-codes
+        names=["d","u","s","c","b","t","g"]
+        pdgs=[1,2,3,4,5,6,21]
+        
+        #the matrix element needs to be initialised
+        matrix2py.initialisemodel(BASE_DIR+"/integrands/mg/"+process+"/param/param_card.dat")
+        
+        #extracting the number of in- and outgoing particles and the masses
+        file = open(BASE_DIR+"/integrands/mg/"+process+"/param/nexternal.inc", "r") 
+        file.readline()
+
```

Comment:
For such parsing instructions that rely on the expected structure of the file, explain what line you are reading, what its structure is and what information you are extracting. I will tag all such instances with [parsing]